### PR TITLE
refactor: Prepare performance aware triggers

### PR DESCRIPTION
**Pull request by robinholzi**
# Motivation

To reuse logic and have a more organised triggering setup, I'd like to make the following adjustments:

# Changes

- Moving core evaluation logic out of `pytorch` evaluator so it's reusable for performance-trigger evaluations.
- drop unused resultwriters
- move `get_embeddings` and `TriggerDatasets` to a more central place as they will be reuse for performance aware triggers.

I already made some comments in https://github.com/eth-easl/modyn/pull/586 that explain what is moved where.

### CODE REVIEW

File: modyn/evaluator/internal/metric_factory.py [hunk 1]
```diff
@@ -1,23 +1,45 @@
...
-    def prepare_metrics(metrics: list[AbstractEvaluationMetric]) -> bool:
-        contains_holistic = False
-        for metric in metrics:
+    def prepare_metrics(metrics: dict[str, AbstractEvaluationMetric]) -> bool:
+        for metric in metrics.values():
             metric.deserialize_evaluation_transformer()
             if isinstance(metric, AbstractHolisticMetric):
                 return True
-        return contains_holistic
+        return False
```
Comment (MaxiBoether) [inline modyn/evaluator/internal/metric_factory.py:45]: This will return true as soon as we reach a holistic metric. This means for metrics after that we won't call the deserialize function. I think we should instead change return True to contains_holistic = True and return that at the end

File: modyn/supervisor/internal/triggers/drift/embedding/embeddings.py [hunk 2]
```diff
@@ -0,0 +1,40 @@
+import torch
+from torch.utils.data import DataLoader
+
+from modyn.supervisor.internal.triggers.utils.model.manager import ModelManager
+
+
+def get_embeddings(model_manager: ModelManager, dataloader: DataLoader) -> torch.Tensor:
```
Comment (MaxiBoether) [inline modyn/supervisor/internal/triggers/drift/embedding/embeddings.py:7]: This needs a test

File: modyn/supervisor/internal/triggers/utils/model/manager.py [hunk 3]
```diff
@@ -11,10 +11,10 @@
 logger = logging.getLogger(__name__)
 
 
-class EmbeddingEncoder:
-    """The EmbeddingEncoder stores a model and its metadata.
+class ModelManager:
```
Comment (MaxiBoether) [inline modyn/supervisor/internal/triggers/utils/model/manager.py:14]: can we find a better term than manager? it's actually considered to be a bit of a code smell to call something manager. maybe there is a gang of four pattern that matches this?

File: modyn/supervisor/internal/triggers/drift/utils.py [hunk 4]
```diff
@@ -0,0 +1,10 @@
+import pandas as pd
+import torch
+
+
+def convert_tensor_to_df(t: torch.Tensor, column_name_prefix: str | None = None) -> pd.DataFrame:
```
Comment (MaxiBoether) [inline modyn/supervisor/internal/triggers/drift/utils.py:5]: This function is missing a test

**Review by MaxiBoether — CHANGES_REQUESTED** [review]
Thank you, this generally looks good but i had a few nitpicky comments

File: modyn/supervisor/internal/triggers/drift/embedding/embeddings.py (hunk 2 above)
Comment (robinholzi) [reply modyn/supervisor/internal/triggers/drift/embedding/embeddings.py:7]: Sure, I'll add one. This function also didn't have one before, though😅

File: modyn/supervisor/internal/triggers/utils/model/manager.py (hunk 3 above)
Comment (robinholzi) [reply modyn/supervisor/internal/triggers/utils/model/manager.py:14]: `Wrapper`? `EmbeddingEncoder` was definitely off, as the class doesn't have anything to do with either embeddings or encoding😅

File: modyn/supervisor/internal/triggers/drift/utils.py (hunk 4 above)
Comment (robinholzi) [reply modyn/supervisor/internal/triggers/drift/utils.py:5]: again technical debt🥶😅

File: modyn/evaluator/internal/metric_factory.py (hunk 1 above)
Comment (robinholzi) [reply modyn/evaluator/internal/metric_factory.py:45]: good catch!

**Review by MaxiBoether — APPROVED** [review]
Thanks. Sorry, i didn't realize those functions were not tested before...

File: modyn/evaluator/internal/metric_factory.py [hunk 5]
```diff
@@ -1,23 +1,46 @@
...
     @staticmethod
-    def prepare_metrics(metrics: list[AbstractEvaluationMetric]) -> bool:
+    def prepare_metrics(metrics: dict[str, AbstractEvaluationMetric]) -> bool:
         contains_holistic = False
-        for metric in metrics:
+        for metric in metrics.values():
             metric.deserialize_evaluation_transformer()
             if isinstance(metric, AbstractHolisticMetric):
-                return True
+                contains_holistic = True
```
Comment (XianzheMa) [inline modyn/evaluator/internal/metric_factory.py:45]: thanks for catching this
